clash-lib/src/Clash/Netlist/Util.hs

```diff
@@ -788,90 +802,122 @@ mkUniqueNormalized is0 topMM (args,binds,res) = do
+        binds1 = binds0 <> maybeToList extraBind
+
       -- Return the uniquely named arguments, let-binders, and result
-      return ( wereVoids
-             , iports
-             , iwrappers
-             , oports
-             , owrappers
-             , zip (bndrsL1 ++ r:bndrsR1) exprs1 ++ extraBndr
-             , Just res1)
+      return (wereVoids, iports, iwrappers, oports, owrappers, binds1, Just res1)
+
     Nothing -> do
       (bndrs1, substArgs1) <- mkUnique substArgs bndrs
-      return ( wereVoids
-             , iports
-             , iwrappers
-             , []
-             , []
-             , zip bndrs1
-                   (map (substTm ("mkUniqueNormalized2" :: Doc ()) substArgs1) exprs)
-             ,Nothing)
-
--- | Set the name of the binder
---
--- Normally, it just keeps the existing name, but there are two exceptions:
---
--- 1. It's the binding for the result which is also referenced by another binding;
---    in this case it's suffixed with `_rec`
--- 2. The binding binds a primitive that has a name control field
+      let binds1 = zip bndrs1 (map (substTm "mkUniqueNormalized2" substArgs1) exprs)
+      return (wereVoids, iports, iwrappers, [], [], binds1, Nothing)
+
+-- | Produce a 'Just' when predicate is True, else Nothing
+orNothing :: Bool -> a -> Maybe a
+orNothing True a = Just a
+orNothing False _ = Nothing
+
+-- | Set the name of the binder if the given term is a blackbox requesting
+-- a specific name for the result binder. It might return multiple names in
+-- case of a multi result primitive.
 --
--- 2. takes priority over 1. Additionally, we create an additional binder when
--- the return value gets a new name.
-setBinderName
-  :: Subst
-  -- ^ Current substitution
-  -> Id
-  -- ^ The binder for the result
-  -> Bool
-  -- ^ Whether the result binder is referenced by another binder
-  -> (Id, Subst, [(Id,Term)])
-  -- ^ * The (renamed) binder for the result
-  --   * The updated substitution in case the result binder is renamed
-  --   * A new binding, to assign the result in case the original binder for
-  --     the result got renamed.
-  -> (Id,Term)
-  -- ^ The binding
-  -> NetlistMonad ((Id, Subst, [(Id,Term)]),Id)
-setBinderName subst res resRead m@(resN,_,_) (i,collectArgsTicks -> (k,args,ticks)) = case k of
-  Prim p -> let nm = primName p in extractPrimWarnOrFail nm >>= go nm
-  _ -> goDef
+renameBinder :: (Id, Term) -> NetlistMonad [(Id, Id)]
+renameBinder (i, collectArgsTicks -> (k, args, ticks)) = withTicks ticks $ \_ -> do
+  case k of
+    Prim p ->
+      case primMultiResult p of
+        SingleResult -> extractPrimWarnOrFail (primName p) >>= goSingle p
+        MultiResult -> extractPrimWarnOrFail (primName p) >>= goMulti p
+    _ -> pure []
  where
-  go nm (BlackBox {resultName = Just (BBTemplate nmD)}) = withTicks ticks $ \_ -> do
-    (bbCtx,_) <- preserveVarEnv (mkBlackBoxContext nm i args)
+  -- Routine for multi result primitives (the default kind of primitive). For
```

Comment:
What do you mean by "the default kind of primitive"?

Reply:
It should have said that for 'single result primitives', not multi result ones!